README.rst
=================
**libaescprintf**
=================

:author: Eddie Brüggemann <mrcyberfighter@gmail.com>

:version: 1.0.0

:License: LGPL

------------
Introduction
------------

**libaescprintf** is a little library based on the ``ANSI escape sequences`` and so only compatible

with ``ANSI escape sequences`` *compliant terminals*.

Most of the terminals that I have tested works well but not all.

In addition to the **foreground** and **background** coloring,

**libaescprintf** provides some styles features like bold, strike-trough, italic, and so on...

-----------------------------
Outputting fancy textual text
-----------------------------

You can use **libaescprintf** for this task in 2 ways:

1. You make usage of the functions from **libaescprintf**.

    They are divided into **2 categories**:

    * Functions which **output** simply to the **standard output stream**: ``printf`` family.

    * Function which **output** to the wanted **stream**: fprintf family.

2. You can use the defines from **libaescprintf** formatting the output using any ``libc`` function.

:note: The goal of this library is to provide a *lightweight textual pretty output*.

-------
Credits
-------

Thanks to my mother, my family and to the doctors.

Stay away from drugs: drugs destroy your brain and your life.

----
NOTE
----

In fact this library is a manner for asking to help for building

a universal lightweight pretty print library using

termcap for getting universal escape.

Because I have try to find it alone without success.

Thanks for any help.

-------
LICENSE
-------

::

  libaescprintf a pretty print library using ANSI escape sequences.
  Copyright (C) 2017  Brüggemann Eddie.

  This library is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published
  by the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 


libaescprintf-1.0.0/NEWS
First release.

I hope to find collaborators for termcap support.

So that we can get a simple little library for formating the output

according to the terminal capabilities.

Not a chunk library like ncurses.


libaescprintf-1.0.0/src/aescprintf.h
/** libaescprintf: ANSI Escape Sequences Color Print Format for styled ouput.
  * Copyright (C) 2017 Brüggemannn Eddie.
  *
  * This library is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Lesser General Public License as published
  * by the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.

  * This library is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  *
  ***************************************************************************/
 
#ifndef LIBAESCPRINTF_1_0_0_HH /** Inclusion guard **/
#define LIBAESCPRINTF_1_0_0_HH

//#include "../config.h"

#if 0
#ifndef _GNU_SOURCE
#define _GNU_SOURCE   // Make an alternative without vasprintf
#endif
#endif

#include <stdio.h>    // For vasprintf
#include <stdarg.h>   // For va_arg, va_copy, va_end, va_list, va_start
#include <stdbool.h>  // For bool type.
#include <string.h>   // for strlen, strcpy,
#include <stdlib.h>   // for free and malloc

// Reading this file will be useful:
#include "./aescprintf_defines.h"

/* Définition of macro EXTERN_C for C++ compatibility */
#ifndef EXTERN_C
#ifdef __cplusplus
#define EXTERN_C extern "C"
#else
#define EXTERN_C extern
#endif
#endif



/** libaescprintf: ANSI Escape Sequences Color printf
  *
  * A mini color printing library based on ANSI escape sequences (instead of termcap.h unfortunately)
  *
  * It should work on most terminals,
  * it work fully on xterm and ANSI escape sequences
  * compliant terminals as terminals build with libvte-2.91.
  *
  ***************************************************************************************************/

#define AESCPRINTF_VERSION_MAJOR (1)
#define AESCPRINTF_VERSION_MINOR (0)
#define AESCPRINTF_VERSION_MICRO (0)


// Style Foreground Background FILE printf
EXTERN_C int scfgbg_fprintf(FILE *fp, style_t style, fg_color_t color, bg_color_t bgcolor, const char *fmt, ...) ;

// Color Foreground Background FILE printf
EXTERN_C int cfgbg_fprintf(FILE *fp, fg_color_t color, bg_color_t bgcolor, const char *fmt, ...) ;

// Color foreground FILE printf
EXTERN_C int cfg_fprintf(FILE *fp, fg_color_t color, const char *fmt, ...) ;

// Color Background FILE printf
EXTERN_C int cbg_fprintf(FILE *fp, bg_color_t bgcolor, const char *fmt, ...) ;

// Style FILE printf
EXTERN_C int s_fprintf(FILE *fp, style_t style, const char *fmt, ...) ;

// Style Color Foreground FILE printf
EXTERN_C int scfg_fprintf(FILE *fp, style_t style, fg_color_t fg_color, const char *fmt, ...) ;

// Style Color Background FILE printf
EXTERN_C int scbg_fprintf(FILE *fp, style_t style, bg_color_t bg_color, const char *fmt, ...) ;

/** printf versions **/

// Style Color Foreground Background printf
EXTERN_C int scfgbg_printf(style_t style, fg_color_t fg_color, bg_color_t bg_color, const char *fmt, ...) ;

// Color Foreground Background printf
EXTERN_C int cfgbg_printf(fg_color_t fg_color, bg_color_t bg_color, const char *fmt, ...) ;

// Color foreground printf
EXTERN_C int cfg_printf(fg_color_t fg_color, const char *fmt, ...) ;

// Color Background printf
EXTERN_C int cbg_printf(bg_color_t bg_color, const char *fmt, ...) ;

// Style printf
EXTERN_C int s_printf(style_t style, const char *fmt, ...) ;

// Style Color Foreground printf
EXTERN_C int scfg_printf(style_t style, fg_color_t fg_color, const char *fmt, ...) ;

EXTERN_C int scbg_printf(style_t style, bg_color_t bg_color, const char *fmt, ...) ;

#endif /** Inclusion guard **/


libaescprintf-1.0.0/src/libaescprintf.c
/** libaescprintf: ANSI Escape Sequences Color Print Format for styled ouput.
  * Copyright (C) 2017 Brüggemannn Eddie.
  *
  * This library is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Lesser General Public License as published
  * by the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.

  * This library is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  *
  ***************************************************************************/

#include "./aescprintf.h"

// Style Color Foreground Background FILE Print Format
int scfgbg_fprintf(FILE *fp, style_t style, fg_color_t fg_color, bg_color_t bg_color, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  if ( ( (fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  if ( ( (bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;
  va_start(aq, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  fprintf(fp, ESC "[%02d;%02d;%02dm",  style, bg_color, fg_color) ;

  /** Else the problem is that the terminal print the background color until the line end.
    * Simply resetting doesn't solve the problem
    */

  char *fmt_str = strdup(fmt) ;

  bool switcher = false ;

  if (fmt_str[strlen(fmt)-1] == '\n') {

    // Construct format string without linefeed.
    fmt_str[strlen(fmt)-1]='\0' ;

    switcher = true ;

  }

  // Print format string without linefeed.
  int ret = vfprintf(fp, fmt_str, ap) ;


  free(fmt_str) ;


  va_end(ap) ;
  va_end(aq) ;



  fprintf(fp, RESET) ;


  if (switcher) {
    // Reset linefeed.
    fprintf(fp,"%c",'\n') ;
  }


  return ret ;

}

// Style Color Foreground Background Print Format
int scfgbg_printf(style_t style, fg_color_t fg_color, bg_color_t bg_color, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  if ( ( (fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  if ( ( (bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  printf(ESC "[%02d;%02d;%02dm",  style, bg_color, fg_color) ;

  /** Else the problem is that the terminal print the background color until the line end.
    * Simply resetting doesn't solve the problem
    */

  char *fmt_str = strdup(fmt) ;

  bool switcher = false ;

  if (fmt_str[strlen(fmt)-1] == '\n') {

    // Construct format string without linefeed.
    fmt_str[strlen(fmt)-1]='\0' ;

    switcher = true ;

  }

  // Print format string without linefeed.
  int ret = vprintf(fmt_str, ap) ;


  free(fmt_str) ;


  va_end(ap) ;
  va_end(aq) ;



  printf(RESET) ;


  if (switcher) {
    // Reset linefeed.
    printf("%c",'\n') ;
  }


  return ret ;

}

// Color Foreground Background FILE Print Format
int cfgbg_fprintf(FILE *fp, fg_color_t fg_color, bg_color_t bg_color, const char *fmt, ...) {

  if ( ((fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  if ( ((bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) )  {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  fprintf(fp, ESC "[%02d;%02dm", bg_color, fg_color) ;

  /** Else the problem is that the terminal print the background color until the line end.
    * Simply resetting doesn't solve the problem
    */

  char *fmt_str = strdup(fmt) ;

  bool switcher = false ;

  if (fmt_str[strlen(fmt)-1] == '\n') {

    // Construct format string without linefeed.
    fmt_str[strlen(fmt)-1]='\0' ;

    switcher = true ;

  }

  // Print format string without linefeed.
  int ret = vfprintf(fp, fmt_str, ap) ;


  free(fmt_str) ;


  va_end(ap) ;
  va_end(aq) ;



  fprintf(fp, RESET) ;


  if (switcher) {
    // Reset linefeed.
    fprintf(fp,"%c",'\n') ;
  }


  return ret ;

}

// Color Foreground Background Print Format
int cfgbg_printf(fg_color_t fg_color, bg_color_t bg_color, const char *fmt, ...) {

  if ( ((fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  if ( ((bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) )  {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  printf(ESC "[%02d;%02dm", bg_color, fg_color) ;

  /** Else the problem is that the terminal print the background color until the line end.
    * Simply resetting doesn't solve the problem
    */

  char *fmt_str = strdup(fmt) ;

  bool switcher = false ;

  if (fmt_str[strlen(fmt)-1] == '\n') {

    // Construct format string without linefeed.
    fmt_str[strlen(fmt)-1]='\0' ;

    switcher = true ;

  }

  // Print format string without linefeed.
  int ret = vprintf(fmt_str, ap) ;


  free(fmt_str) ;


  va_end(ap) ;
  va_end(aq) ;



  printf(RESET) ;


  if (switcher) {
    // Reset linefeed.
    printf("%c",'\n') ;
  }


  return ret ;

}


// Color foreground FILE Print Format
int cfg_fprintf(FILE *fp, fg_color_t fg_color, const char *fmt, ...) {

  if ( ((fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }



  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  fprintf(fp, ESC "[%01d;%02dm", normal, fg_color) ;

  int ret = vfprintf(fp, fmt, ap) ;

  va_end(ap) ;
  va_end(aq) ;

  fprintf(fp, RESET) ;

  return ret ;

}

// Color foreground Print Format
int cfg_printf(fg_color_t fg_color, const char *fmt, ...) {

  if ( ((fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  printf(ESC "[%01d;%02dm", normal, fg_color) ;

  int ret = vprintf(fmt, ap) ;

  va_end(ap) ;
  va_end(aq) ;

  printf(RESET) ;

  return ret ;

}

// Color Background FILE Print Format
int cbg_fprintf(FILE *fp, bg_color_t bg_color, const char *fmt, ...) {


  if ( ((bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) )  {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  fprintf(fp, ESC "[%01d;%02dm", normal, bg_color) ;

  /** Else the problem is that the terminal print the background color until the line end.
    * Simply resetting doesn't solve the problem
    */

  char *fmt_str = strdup(fmt) ;

  bool switcher = false ;

  if (fmt_str[strlen(fmt)-1] == '\n') {

    // Construct format string without linefeed.
    fmt_str[strlen(fmt)-1]='\0' ;

    switcher = true ;

  }

  // Print format string without linefeed.
  int ret = vfprintf(fp, fmt_str, ap) ;


  free(fmt_str) ;


  va_end(ap) ;
  va_end(aq) ;



  fprintf(fp, RESET) ;


  if (switcher) {
    // Reset linefeed.
    fprintf(fp,"%c",'\n') ;
  }


  return ret ;

}

// Style Color Background FILE Print Format
int scbg_fprintf(FILE *fp, style_t style, bg_color_t bg_color, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  if ( ((bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) )  {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs



  fprintf(fp, ESC "[%02d;%02dm", style, bg_color) ;


  int ret = vfprintf(fp, fmt, ap) ;


  va_end(ap) ;
  va_end(aq) ;

  fprintf(fp, RESET) ;

  return ret ;

}

// Color Background Print Format
int cbg_printf(bg_color_t bg_color, const char *fmt, ...) {


  if ( ((bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) )  {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  printf(ESC "[%01d;%02dm", normal, bg_color) ;
  printf(ESC "[%01d;%02dm", normal, bg_color) ;
  /** Else the problem is that the terminal print the background color until the line end.
    * Simply resetting doesn't solve the problem
    */

  char *fmt_str = strdup(fmt) ;

  bool switcher = false ;

  if (fmt_str[strlen(fmt)-1] == '\n') {

    // Construct format string without linefeed.
    fmt_str[strlen(fmt)-1]='\0' ;

    switcher = true ;

  }

  // Print format string without linefeed.
  int ret = vprintf(fmt_str, ap) ;


  free(fmt_str) ;


  va_end(ap) ;
  va_end(aq) ;



  printf(RESET) ;


  if (switcher) {
    // Reset linefeed.
    printf("%c",'\n') ;
  }


  return ret ;

}

// Style FILE Print Format
int s_fprintf(FILE *fp, style_t style, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

  fprintf(fp, ESC "[%01dm", style) ;

  int ret = vfprintf(fp, fmt, ap) ;

  va_end(ap) ;
  va_end(aq) ;

  fprintf(fp, RESET) ;

  return ret ;

}

// Style Print Format
int s_printf(style_t style, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one need.

  printf(ESC "[%01dm", style) ;

  int ret = vprintf(fmt, ap) ;

  va_end(ap) ;
  va_end(aq) ;

  printf(RESET) ;

  return ret ;

}

// Style Color Foreground FILE Print Format
int scfg_fprintf(FILE *fp, style_t style, fg_color_t fg_color, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  if ( ( (fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  arg_err_msg(fp) ;

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs



  fprintf(fp, ESC "[%02d;%02dm", style, fg_color) ;


  int ret = vfprintf(fp, fmt, ap) ;


  va_end(ap) ;
  va_end(aq) ;

  fprintf(fp, RESET) ;

  return ret ;

}

// Style Color Foreground Print Format
int scfg_printf(style_t style, fg_color_t fg_color, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  if ( ( (fg_color < 30 || fg_color > 39) || fg_color == 38 ) && (fg_color < 90 || fg_color > 97) ) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument fg_color must be from type fg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs





  printf(ESC "[%02d;%02dm", style, fg_color) ;


  int ret = vprintf(fmt, ap) ;


  va_end(ap) ;
  va_end(aq) ;

  printf(RESET) ;

  return ret ;

}

// Style Color Foreground Print Format
int scbg_printf(style_t style, bg_color_t bg_color, const char *fmt, ...) {

  if (style < 0 || style > 9) {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument style must be from type style_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }


  if ( ((bg_color < 30+10 || bg_color > 39+10) || bg_color == 38+10) && (bg_color < 90+10 || bg_color > 97+10) )  {

    fprintf(stderr, RESET) ;

    fprintf(stderr,"%s:%s:%d\n" "argument bg_color must be from type bg_color_t !\n", __FILE__, __func__, __LINE__) ;

    return -1 ;
  }

  va_list ap, aq ;

  va_start(ap, fmt) ;

  va_copy(aq, ap); // __va_copy(aq, ap) ;
  // *aq = *ap; // Unfortunately, there are also systems that make it an array of pointers (of length 1), and there one needs

   printf(ESC "[%02d;%02dm", style, bg_color) ;


  int ret = vprintf(fmt, ap) ;


  va_end(ap) ;
  va_end(aq) ;

  printf(RESET) ;

  return ret ;

}



libaescprintf-1.0.0/examples/examples_matrix.c
/** libaescprintf: ANSI Escape Sequences Color Print Format for styled ouput.
  * Copyright (C) 2017 Brüggemannn Eddie.
  *
  * This library is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Lesser General Public License as published
  * by the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.

  * This library is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  *
  ***************************************************************************/

#include <aescprintf.h>

#include <time.h>

void matrix(void) {

  /** Print a matrix in a terminal using @libaescprintf **/

  fg_color_t fg_colors[] = { fg_blue, fg_cyan, fg_darkgray, fg_green, fg_green, fg_lightblue, fg_lightcyan, fg_lightgray, fg_lightgreen, fg_lightpink, fg_lightred, fg_lightred, fg_lightyellow } ;

  unsigned int counter = 0 ;

  srand(time(NULL)) ;

  while (counter < 4294967295-1)  {

    char *c = (char *) malloc(sizeof(char)) ;

    *c = 48 ;

    *c += rand() % 10 ;

    cfg_printf(fg_colors[rand() % sizeof(fg_colors)/sizeof(int)], "%c", *c) ;



    fflush(stdout) ;

    free(c) ;

 
    ++counter ;
 

  }

}


int main(void) {

  matrix() ;

  return 0 ;

}



libaescprintf-1.0.0/src/aescprintf_defines.h
/** libaescprintf: ANSI Escape Sequences Color Print Format for styled ouput.
  * Copyright (C) 2017 Brüggemannn Eddie.
  *
  * This library is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Lesser General Public License as published
  * by the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.

  * This library is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  *
  ***************************************************************************/

#ifndef LIBAESCPRINTF_1_0_0_DEFINES_HH /** Inclusion guard **/
#define LIBAESCPRINTF_1_0_0_DEFINES_HH

/** About definition: **/
#define AUTHOR      "Brüggemann Eddie"
#define MAIL        "<mrcyberfighter@gmail.com>"
#define LIBNAME     "libaescprintf"
#define VERSION     "1.0.0"
#define LICENSE     "lgpl"
#define PRGWEBURL   ""
#define COPYRIGHT   "Copyright (C) 2017 Brüggemann Eddie"

/** libaescprintf: Ansi Escape Sequences Color printf
  *
  * A mini color printing library based on ANSI escape sequence (instead of termcap.h)
  *
  * It should work on most terminals, it work fully on xterm.
  *
  */


// Building a color and a style type. (Used into library functions).

enum style_enum { normal = 0 , bold = 1 , dim = 2 , italic = 3, underline = 4 , blink = 5, inverted = 7, hidden = 8, strikethrough = 9 } ;

//typedef enum color_enum  color_t ;
typedef enum style_enum  style_t ;

// Building a foreground color type. (Used into library functions).

                     // Foreground Normal >= 30   <= 37   (8 colors)
                     // Foreground Light  >= 90   <= 97   (8 colors)
enum fg_color_enum { fg_black     = 30,     fg_darkgray     = 90,
                     fg_red       = 31,     fg_lightred     = 91,
                     fg_green     = 32,     fg_lightgreen   = 92,
                     fg_yellow    = 33,     fg_lightyellow  = 93,
                     fg_blue      = 34,     fg_lightblue    = 94,
                     fg_pink      = 35,     fg_lightpink    = 95,
                     fg_cyan      = 36,     fg_lightcyan    = 96,
                     fg_lightgray = 37,     fg_white        = 97,
          
                     fg_default   = 39 } ;

                  
// Building a background color type. (Used into library functions).                  
                  
                     // Background Normal >= 40   <= 47   (8 colors)
                     // Background Light  >= 100  <= 107  (8 colors)
enum bg_color_enum { bg_black     = (30+10),     bg_darkgray    = (90+10),
                     bg_red       = (31+10),     bg_lightred    = (91+10),
                     bg_green     = (32+10),     bg_lightgreen  = (92+10),
                     bg_yellow    = (33+10),     bg_lightyellow = (93+10),
                     bg_blue      = (34+10),     bg_lightblue   = (94+10),
                     bg_pink      = (35+10),     bg_lightpink   = (95+10),
                     bg_cyan      = (36+10),     bg_lightcyan   = (96+10),
                     bg_lightgray = (37+10),     bg_white       = (97+10),
          
                     bg_default   = (39+10) } ;

typedef enum fg_color_enum  fg_color_t ;
typedef enum bg_color_enum  bg_color_t ;


/** You can use this defines into raw string formatting. **/

#define ESC   "\033"  // Escape sequence.

// Dark colors defines.
#define DB    ESC "[30m"  // Dark Black.
#define DR    ESC "[31m"  // Dark Red.
#define DG    ESC "[32m"  // Dark Green.
#define DY    ESC "[33m"  // Dark Yellow.
#define DS    ESC "[34m"  // Dark Sky. (Dark blue)
#define DP    ESC "[35m"  // Dark Pink.
#define DC    ESC "[36m"  // Dark Cyan.
#define DW    ESC "[37m"  // Dark white. (Light Gray)

// Light colors defines.
#define LB    ESC "[90m"  // Light Black.
#define LR    ESC "[91m"  // Light Red.
#define LG    ESC "[92m"  // Light Green.
#define LY    ESC "[93m"  // Light Yellow.
#define LS    ESC "[94m"  // Light Sky.  (Light blue)
#define LP    ESC "[95m"  // Light Pink.
#define LC    ESC "[96m"  // Light Cyan.
#define LW    ESC "[97m"  // Light White.

#define DD    ESC "[39m"  // Default color.

// Styles defines
#define RESET           ESC "[0m"   // Reset.
#define BOLD            ESC "[1m"   // Bold.
#define DIM             ESC "[2m"   // DIM
#define ITALIC          ESC "[3m"   // Italic.
#define UNDERSTRIKE     ESC "[4m"   // Under-strike.
#define BLINK           ESC "[5m"   // Blink (Doesn't work on all terminals look at xterm it work's on it !).
#define INVERTED        ESC "[7m"   // Inverted (Invert foreground and background).
#define HIDDEN          ESC "[8m"   // Hidden (Print only background color).
#define STRIKETROUGHT   ESC "[9m"   // Strike-through.


#ifndef NO_REDIRECT
#define arg_err_msg(fp) if (fp != stdout && fp != stderr) { fprintf(stderr,"%s(...) argument fp must be: 'stdout' or 'stderr'\n", __func__) ; fflush(stderr) ; return -1 ; }
#endif


#endif /** Inclusion guard **/
